test/factories/host_related.rb

```diff
@@ -1,13 +1,83 @@
+    sequence(:value) { |n| "parameter value #{n}" }
+    type 'CommonParameter'
+  end
+  factory :host_parameter, :parent => :parameter, :class => HostParameter do
+    type 'HostParameter'
+  end
+  factory :hostgroup_parameter, :parent => :parameter, :class => GroupParameter do
+    type 'GroupParameter'
+  end
+  factory :nic_base do
+    sequence(:identifier) { |n| "eth#{n}" }
+    sequence(:mac) { |n| "00:00:00:00:" + n.to_s(16).rjust(4, '0').insert(2, ':') }
+  end
+  factory :nic_interface, :class => Nic::Interface, :parent => :nic_base do
+    type 'Nic::Interface'
+  end
+  factory :nic_managed, :class => Nic::Managed, :parent => :nic_interface do
+    type 'Nic::Managed'
+    sequence(:mac) { |n| "01:23:45:ab:" + n.to_s(16).rjust(4, '0').insert(2, ':') }
+    sequence(:ip) { |n| IPAddr.new(n, Socket::AF_INET).to_s }
+  end
+  factory :nic_bmc, :class => Nic::BMC, :parent => :nic_managed do
+    type 'Nic::BMC'
+    sequence(:mac) { |n| "01:23:56:cd:" + n.to_s(16).rjust(4, '0').insert(2, ':') }
+    sequence(:ip) { |n| IPAddr.new(255 + n, Socket::AF_INET).to_s }
+    provider 'IPMI'
+    username 'admin'
+    password 'admin'
+  end
+  factory :nic_bond, :class => Nic::Bond, :parent => :nic_managed do
+    type 'Nic::Bond'
+    mode 'balance-rr'
+  end
+  factory :nic_primary_and_provision, :parent => :nic_managed, :class => Nic::Managed do
+    primary true
+    provision true
+    sequence(:mac) { |n| "01:23:45:ab:" + n.to_s(16).rjust(4, '0').insert(2, ':') }
+    sequence(:ip) { |n| IPAddr.new(n, Socket::AF_INET).to_s }
+  end
+
   factory :host do
     sequence(:name) { |n| "host#{n}" }
     sequence(:hostname) { |n| "host#{n}" }
-    sequence(:ip) { |n| IPAddr.new(n, Socket::AF_INET).to_s }
-    sequence(:mac) { |n| "01:23:45:67:" + n.to_s(16).rjust(4, '0').insert(2, ':') }
     root_pass 'xybxa6JUkz63w'
-    domain
     environment
 
+    # This allows a test to declare build/create(:host, :ip => '1.2.3.4') and
+    # have the primary interface correctly updated with the specified attrs
+    ignore do
+      ip     { false }
+      mac    { false }
+      subnet { false }
+      domain { false }
+    end
+    after(:build) do |host,evaluator|
+      if host.primary_interface
+        host.primary_interface.mac     = evaluator.mac    if evaluator.mac
+        host.primary_interface.ip      = evaluator.ip     if evaluator.ip
+        host.primary_interface.subnet  = evaluator.subnet if evaluator.subnet
+        host.primary_interface.domain  = evaluator.domain if evaluator.domain
+      else
+        opts = { :primary => true,
+                 :domain  => FactoryGirl.build(:domain)
+        }
+        opts[:mac]    = evaluator.mac    if evaluator.mac
+        opts[:ip]     = evaluator.ip     if evaluator.ip
+        opts[:subnet] = evaluator.subnet if evaluator.subnet
+        opts[:domain] = evaluator.domain if evaluator.domain
+        host.interfaces << FactoryGirl.build(:nic_managed, opts)
+      end
```

Comment:
Lots of duplicated code in the ```after(:build)``` block in my opinion, how about a refactor like..

```ruby
nic_object = if host.primary_interface.present? 
                      host.primary_interface
                    else
                     { :primary => true, :domain => FactoryGirl.build(:domain) }
                    end

set_nic_attributes(nic_object, [:mac, :ip, :subnet, :domain]) 

def set_nic_attributes(nic_object, attributes)
  attributes.each do |nic_attribute|               
    next unless evaluator.send(nic_attribute).present?
    nic_object.send(:[]=, nic_attribute, evaluator.send(nic_attribute))
  end
  nic_object
end
```

Current code will be hard to maintain if we happen to add more attributes to interface. Same for the ```ignore``` block above, if we could have a single place to define all the NIC attributes we want to ignore and set after build it'd be great.